File: agents/myrmec/agent/http_client.py
"""
HTTP client for agent registration and token refresh.
"""

import socket
from typing import Any
from uuid import UUID

import httpx
from pydantic import BaseModel


class RegisterRequest(BaseModel):
    """Registration request payload."""
    hostname: str
    ip_address: str | None = None
    sdk_version: str
    metadata: dict[str, Any] | None = None


class RegisterResponse(BaseModel):
    """Registration response from Engine."""
    access_token: str
    refresh_token: str
    instance_id: UUID


class RefreshResponse(BaseModel):
    """Token refresh response from Engine."""
    access_token: str
    refresh_token: str


class RetrievalHit(BaseModel):
    """A single retrieval hit returned by ``POST /api/v1/agent/retrieve``."""

    passage: str
    chunk_id: UUID
    source_id: UUID
    source_name: str
    locator: str
    score: float

    @classmethod
    def from_payload(cls, payload: dict[str, Any]) -> "RetrievalHit":
        """Build a hit from the raw engine JSON (camelCase keys)."""
        return cls(
            passage=payload["passage"],
            chunk_id=UUID(payload["chunkId"]),
            source_id=UUID(payload["sourceId"]),
            source_name=payload["sourceName"],
            locator=payload["locator"],
            score=float(payload["score"]),
        )



class EngineHttpClient:
    """
    HTTP client for agent registration and token management.
    
    Handles:
    - Initial registration with registration key
    - Token refresh using refresh token
    """
    
    def __init__(
        self,
        engine_url: str,
        registration_key: str,
        sdk_version: str = "0.1.0",
        timeout: float = 30.0,
    ):
        """
        Initialize HTTP client.
        
        Args:
            engine_url: Base URL of the Engine (e.g., "http://localhost:8080").
            registration_key: Agent registration key (myr_agent_...).
            sdk_version: SDK version to report.
            timeout: HTTP request timeout in seconds.
        """
        self._engine_url = engine_url.rstrip("/")
        self._registration_key = registration_key
        self._sdk_version = sdk_version
        self._timeout = timeout
        
        self._client = httpx.AsyncClient(timeout=timeout)
    
    async def register(
        self,
        metadata: dict[str, Any] | None = None,
    ) -> RegisterResponse:
        """
        Register agent instance with the Engine.
        
        Args:
            metadata: Optional metadata about this instance.
        
        Returns:
            RegisterResponse with access/refresh tokens and instance ID.
        
        Raises:
            httpx.HTTPStatusError: If registration fails.
        """
        hostname = socket.gethostname()
        ip_address = self._get_ip_address()
        
        request = RegisterRequest(
            hostname=hostname,
            ip_address=ip_address,
            sdk_version=self._sdk_version,
            metadata=metadata,
        )
        
        response = await self._client.post(
            f"{self._engine_url}/api/v1/agent/auth/register",
            json=request.model_dump(by_alias=True, exclude_none=True),
            headers={"X-Registration-Key": self._registration_key},
        )
        response.raise_for_status()
        
        data = response.json()
        return RegisterResponse(
            access_token=data["accessToken"],
            refresh_token=data["refreshToken"],
            instance_id=UUID(data["instanceId"]),
        )
    
    async def refresh(self, refresh_token: str) -> RefreshResponse:
        """
        Refresh access token using refresh token.
        
        Args:
            refresh_token: Current refresh token.
        
        Returns:
            RefreshResponse with new access/refresh tokens.
        
        Raises:
            httpx.HTTPStatusError: If refresh fails.
        """
        response = await self._client.post(
            f"{self._engine_url}/api/v1/agent/auth/refresh",
            headers={"Authorization": f"Bearer {refresh_token}"},
        )
        response.raise_for_status()
        
        data = response.json()
        return RefreshResponse(
            access_token=data["accessToken"],
            refresh_token=data["refreshToken"],
        )
    
    async def close(self) -> None:
        """Close the HTTP client."""
        await self._client.aclose()

    async def retrieve(
        self,
        access_token: str,
        knowledge_base_id: UUID | str,
        query: str,
        top_k: int = 5,
        filters: dict[str, str] | None = None,
    ) -> list[RetrievalHit]:
        """
        Run a retrieval query against a knowledge base.

        Backs the ``ctx.retrieve()`` ergonomic helper that ships with the
        chat runtime in a later release. For now agents that want RAG
        grounding can call this directly with their current access token.

        Args:
            access_token: Agent access token (from ``register``/``refresh``).
            knowledge_base_id: Target knowledge base UUID.
            query: Free-text retrieval query.
            top_k: Maximum number of hits to return.
            filters: Optional provider-specific filters.

        Returns:
            Ordered list of ``RetrievalHit`` (highest-score first). Empty
            list if the provider returns nothing or fails — the engine
            surfaces provider failures as an empty 200 so a misbehaving
            knowledge base never aborts an agent run.

        Raises:
            httpx.HTTPStatusError: On 4xx (bad request / auth) responses.
        """
        body: dict[str, Any] = {
            "knowledgeBaseId": str(knowledge_base_id),
            "query": query,
            "topK": top_k,
        }
        if filters:
            body["filters"] = filters

        response = await self._client.post(
            f"{self._engine_url}/api/v1/agent/retrieve",
            json=body,
            headers={"Authorization": f"Bearer {access_token}"},
        )
        response.raise_for_status()
        return [RetrievalHit.from_payload(item) for item in response.json()]

    
    def _get_ip_address(self) -> str | None:
        """Get local IP address."""
        try:
            # Connect to a public DNS to determine local IP
            s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
            s.connect(("8.8.8.8", 80))
            ip = s.getsockname()[0]
            s.close()
            return ip
        except Exception:
            return None
    
    @property
    def engine_url(self) -> str:
        """Engine base URL."""
        return self._engine_url

File: agents/src/transport/httpClient.ts
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 The Myrmec Authors

/**
 * HTTP client for agent registration and token refresh.
 *
 * Ported from the Python SDK (`myrmec/agent/http_client.py` →
 * `EngineHttpClient`). Uses the Node-native `fetch` (Node 18+). Backs the
 * Supervisor's bootstrap (REQ-A-001) and token lifecycle (REQ-A-002).
 *
 * The engine speaks camelCase JSON (project convention); request/response
 * keys here match the wire exactly.
 */
import { hostname } from "node:os";

/** Result of a successful registration. */
export interface RegisterResponse {
  accessToken: string;
  refreshToken: string;
  /** Engine-assigned instance/agent id (UUID string). */
  instanceId: string;
}

/** Result of a successful token refresh. */
export interface RefreshResponse {
  accessToken: string;
  refreshToken: string;
}

/**
 * A single retrieval hit returned by `POST /api/v1/agent/retrieve`.
 *
 * Keys match the engine wire (camelCase) exactly, so hits are used as parsed.
 */
export interface RetrievalHit {
  passage: string;
  /** Chunk UUID (string). */
  chunkId: string;
  /** Source UUID (string). */
  sourceId: string;
  sourceName: string;
  locator: string;
  score: number;
}

/** Inputs for a {@link EngineHttpClient.retrieve} call. */
export interface RetrieveOptions {
  /** Target knowledge base UUID. */
  knowledgeBaseId: string;
  /** Free-text retrieval query. */
  query: string;
  /** Maximum number of hits to return (default 5). */
  topK?: number;
  /** Optional provider-specific filters. */
  filters?: Record<string, string>;
  /**
   * Optional audit context. When supplied, the engine records a `RETRIEVAL`
   * execution event (feature #32) capturing the query and returned
   * chunk/source IDs + scores so an AUDITOR can replay which knowledge fed an
   * answer. Must reference a real workflow task; omit for contextless
   * retrievals.
   */
  taskId?: string;
  /** Optional attempt id paired with {@link taskId}. */
  attemptId?: string;
}

export interface EngineHttpClientOptions {
  /** Base Engine URL, e.g. "http://localhost:8080". */
  engineUrl: string;
  /** Agent registration key (`myr_agent_…`). */
  registrationKey: string;
  /** SDK version reported to the Engine. */
  sdkVersion?: string;
  /** Per-request timeout in ms (default 30 000). */
  timeoutMs?: number;
  /** Optional metadata recorded with the registration. */
  metadata?: Record<string, unknown>;
}

/** Thrown when an Engine HTTP call returns a non-2xx status. */
export class EngineHttpError extends Error {
  constructor(
    readonly status: number,
    readonly path: string,
    readonly body: string,
  ) {
    super(`Engine ${path} failed: HTTP ${status}`);
    this.name = "EngineHttpError";
  }
}

export class EngineHttpClient {
  private readonly engineUrl: string;
  private readonly registrationKey: string;
  private readonly sdkVersion: string;
  private readonly timeoutMs: number;
  private readonly metadata?: Record<string, unknown>;

  constructor(options: EngineHttpClientOptions) {
    this.engineUrl = options.engineUrl.replace(/\/+$/, "");
    this.registrationKey = options.registrationKey;
    this.sdkVersion = options.sdkVersion ?? "0.1.0";
    this.timeoutMs = options.timeoutMs ?? 30_000;
    this.metadata = options.metadata;
  }

  /** Register this host with the Engine; returns the first token pair. */
  async register(): Promise<RegisterResponse> {
    const body: Record<string, unknown> = {
      hostname: hostname(),
      sdkVersion: this.sdkVersion,
    };
    if (this.metadata) {
      body.metadata = this.metadata;
    }

    const data = await this.post(
      "/api/v1/agent/auth/register",
      { "X-Registration-Key": this.registrationKey },
      body,
    );
    return {
      accessToken: data.accessToken as string,
      refreshToken: data.refreshToken as string,
      instanceId: data.instanceId as string,
    };
  }

  /** Exchange a refresh token for a fresh token pair. */
  async refresh(refreshToken: string): Promise<RefreshResponse> {
    const data = await this.post(
      "/api/v1/agent/auth/refresh",
      { Authorization: `Bearer ${refreshToken}` },
      undefined,
    );
    return {
      accessToken: data.accessToken as string,
      refreshToken: data.refreshToken as string,
    };
  }

  /**
   * Run a retrieval query against a knowledge base.
   *
   * Backs the `ctx.retrieve()` ergonomic helper that ships with the chat
   * runtime. Agents that want RAG grounding can also call this directly with
   * their current access token.
   *
   * The engine surfaces provider failures as an empty 200, so a misbehaving
   * knowledge base returns `[]` rather than aborting an agent run; only 4xx
   * (bad request / auth) responses throw {@link EngineHttpError}.
   *
   * @param accessToken Agent access token (from `register`/`refresh`).
   * @param options Retrieval inputs, including optional audit context.
   * @returns Ordered hits (highest-score first); empty when nothing matches.
   */
  async retrieve(
    accessToken: string,
    options: RetrieveOptions,
  ): Promise<RetrievalHit[]> {
    const body: Record<string, unknown> = {
      knowledgeBaseId: options.knowledgeBaseId,
      query: options.query,
      topK: options.topK ?? 5,
    };
    if (options.filters) {
      body.filters = options.filters;
    }
    if (options.taskId !== undefined) {
      body.taskId = options.taskId;
    }
    if (options.attemptId !== undefined) {
      body.attemptId = options.attemptId;
    }

    const data = await this.post<RetrievalHit[]>(
      "/api/v1/agent/retrieve",
      { Authorization: `Bearer ${accessToken}` },
      body,
    );
    return Array.isArray(data) ? data : [];
  }

  /**
   * Fetch the raw bytes of a conversation attachment (#103 Slice A).
   *
   * Backs native image inlining: the agent turns an image attachment's
   * `readContentPath` into a multimodal image part by GETting its bytes here.
   * Non-2xx responses throw {@link EngineHttpError} so the caller can degrade
   * to a text note rather than crash the turn.
   *
   * @param accessToken Agent access token (from `register`/`refresh`).
   * @param path Engine read path (e.g. `/api/v1/agent/conversations/…/content`).
   * @returns The attachment bytes.
   */
  async fetchAttachmentContent(
    accessToken: string,
    path: string,
  ): Promise<Uint8Array> {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeoutMs);
    try {
      const response = await fetch(`${this.engineUrl}${path}`, {
        method: "GET",
        headers: { Authorization: `Bearer ${accessToken}` },
        signal: controller.signal,
      });
      if (!response.ok) {
        const text = await response.text().catch(() => "");
        throw new EngineHttpError(response.status, path, text);
      }
      const buffer = await response.arrayBuffer();
      return new Uint8Array(buffer);
    } finally {
      clearTimeout(timer);
    }
  }

  /** Base URL the client targets. */
  get baseUrl(): string {
    return this.engineUrl;
  }

  private async post<T = Record<string, unknown>>(
    path: string,
    headers: Record<string, string>,
    body: Record<string, unknown> | undefined,
  ): Promise<T> {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeoutMs);
    try {
      const response = await fetch(`${this.engineUrl}${path}`, {
        method: "POST",
        headers: {
          ...headers,
          ...(body ? { "Content-Type": "application/json" } : {}),
        },
        body: body ? JSON.stringify(body) : undefined,
        signal: controller.signal,
      });

      const text = await response.text();
      if (!response.ok) {
        throw new EngineHttpError(response.status, path, text);
      }
      return (text ? JSON.parse(text) : {}) as T;
    } finally {
      clearTimeout(timer);
    }
  }
}
